internal/pkg/v2/application/validator/common/common.go

```diff
@@ -0,0 +1,55 @@
+/*******************************************************************************
+ * Copyright 2020 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package common
+
+import (
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/application"
+	dtoBase "github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/dto/v2dot0/base"
+	dtoError "github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/dto/v2dot0/error"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/validator/base"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/v2/application/validator/behavior"
+)
+
+// getRequestID defines the signature of a closure provided to Validate that is called to obtain the requestID from
+// the caller's type-asserted struct.
+type getRequestID func() string
+
+// Validate handles common generic validation after initial type assertion; includes base validation and behavior
+// validation.
+func Validate(
+	r interface{},
+	assertionOK bool,
+	getRequestID getRequestID,
```

Comment:
Why must this be a function rather than just pass in the requestID

Reply:
Precedence.

Validate function is called immediately after a type assertion intended to convert a request (of type interface{}) to a use-case-specific request DTO.  It's generic code (and therefore has no knowledge of the concrete DTO the request is being asserted into.

Further, the first action of Validate is to test whether the type assertion was successful.

The follow-on validation -- which is generic in nature -- requires the requestID from the request being validated.  There's no other way to pass that into Validate without duplicating the common code to verify the type assertion result.  For example, it isn't safe to do this (from validator/version/version.go)...

```
	request, ok := r.(*dto.Request)
	response, status = common.Validate(
		r,
		ok,
                request.RequestID,
		behavior,
		application.NewBehaviorSlice(application.Version2, application.KindVersion, application.ActionCommand),
	)
```

...because we don't yet know the value of request.  It could be nil -- which would cause a panic.  It's only after we ensure the type assertion succeeded that we can leverage the asserted types instance to access its properties.